processor/src/fast/checkpoints.rs

```diff
@@ -0,0 +1,647 @@
+    }
+
+    /// Records a Merkle path verification with its address and computed root
+    pub fn record_build_merkle_root(&mut self, addr: Felt, computed_root: Word) {
+        self.build_merkle_root_operations.push_back((addr, computed_root));
+    }
+
+    /// Records a Merkle root update with its address, old root, and new root
+    pub fn record_update_merkle_root(&mut self, addr: Felt, old_root: Word, new_root: Word) {
+        self.mrupdate_operations.push_back((addr, old_root, new_root));
+    }
+
+    // ACCESSORS (used by parallel trace generators)
+    // --------------------------------------------------------------------------------
+
+    /// Replays a `Hasher::hash_control_block` or `Hasher::hash_basic_block` operation, returning
+    /// the pre-recorded address
+    pub fn replay_block_address(&mut self) -> Felt {
+        self.block_addresses.pop_front().expect("No block address operations recorded")
+    }
+
+    /// Replays a `Hasher::permute` operation, returning its address and result
+    pub fn replay_permute(&mut self) -> (Felt, [Felt; 12]) {
+        self.permutation_operations
+            .pop_front()
+            .expect("No permutation operations recorded")
+    }
+
+    /// Replays a Merkle path verification, returning the pre-recorded address and computed root
+    pub fn replay_build_merkle_root(&mut self) -> (Felt, Word) {
+        self.build_merkle_root_operations
+            .pop_front()
+            .expect("No build merkle root operations recorded")
+    }
+
+    /// Replays a Merkle root update, returning the pre-recorded address, old root, and new root
+    pub fn replay_update_merkle_root(&mut self) -> (Felt, Word, Word) {
+        self.mrupdate_operations.pop_front().expect("No mrupdate operations recorded")
+    }
+}
+
+/// Implements a shim for stack overflow operations, in which all overflow values and addresses
+/// during a given fragment are pre-recorded by the fast processor and replayed by the main trace
+/// fragment generators.
+///
+/// This is used to simulate stack overflow functionality in parallel trace generation without
+/// needing to maintain the actual overflow table. All overflow operations are recorded during
+/// fast execution and then replayed during parallel trace generation.
+///
+/// The shim records overflow values (from pop operations) and overflow addresses (representing
+/// the clock cycle of the last overflow update) and provides replay methods that return the
+/// pre-recorded values. This works naturally since the fast processor has exactly the same
+/// access patterns as the main trace generators.
+#[derive(Debug)]
+pub struct StackOverflowReplay {
+    /// Recorded overflow values and overflow addresses from pop_overflow operations. Each entry
+    /// represents a value that was popped from the overflow stack, and the overflow address of the
+    /// at the top of the overflow stack *after* the pop operation.
```

Comment:
"...of the at the top..." needs a bit of simplification ?

Reply:
Done - added an example too to be more clear